src/fairseq2/datasets/huggingface.py

```diff
@@ -0,0 +1,192 @@
+    create_bucket_sizes,
+    read_sequence,
+)
+from fairseq2.datasets.data_reader import DataPipelineReader
+from fairseq2.gang import Gang
+from fairseq2.models.seq2seq import Seq2SeqBatch
+
+try:
+    from datasets import (  # type: ignore[attr-defined,import-untyped,import-not-found]
+        Dataset,
+        DatasetDict,
+    )
+except ImportError:
+    has_datasets = False
+else:
+    has_datasets = True
+
+
+Example = Dict[str, Any]
+
+
+class Batcher(abc.ABC):
+    @abc.abstractmethod
+    def batch(self, builder: DataPipelineBuilder, **kwargs: Any) -> DataPipelineBuilder:
+        ...
+
+
+class BatcherBySize(Batcher):
+    def __init__(self, bucket_size: int) -> None:
+        if bucket_size <= 0:
+            raise ValueError(f"Forbidden value: bucket_size={bucket_size}")
+        self.bucket_size = bucket_size
+
+    def batch(self, builder: DataPipelineBuilder, **kwargs: Any) -> DataPipelineBuilder:
+        return builder.bucket(bucket_size=self.bucket_size)
+
+
+class BatcherBySeqLength(Batcher):
+    """
+    The batching strategy that looks into the sequence defined in the column
+    `seq_len_col`, and batch them until certain accumulated length is reached.
+
+    :param max_num_elements:
+        The maximum number of elements in each batch.
+    :param max_seq_len:
+        The maximum sequence length of each example.
+        Examples longer than this value will be cropped.
+    :param min_seq_len:
+        The minimum sequence length of each example. Examples
+        shorter than this value will be dropped.
+    :param seq_len_col:
+        The column that stores length of source sequence. Must
+        be provided if the data is to be batched by length
+    """
+
+    def __init__(
+        self,
+        max_num_elements: int,
+        max_seq_len: Optional[int] = None,
+        min_seq_len: Optional[int] = None,
+        seq_len_col: Optional[str] = None,
+        **extra: Any,
+    ):
+        self.max_num_elements = max_num_elements
+        self.max_seq_len = max_seq_len or max_num_elements
+        self.min_seq_len = min_seq_len or 1
+        self.skip_below_min_examples = min_seq_len is not None
+        self.seq_len_col = seq_len_col
+
+    def batch(
+        self, builder: DataPipelineBuilder, num_seqs_multiple_of: int = 1, **kwargs: Any
+    ) -> DataPipelineBuilder:
+        bucket_sizes = create_bucket_sizes(
+            max_num_elements=self.max_num_elements,
+            max_seq_len=self.max_seq_len,
+            min_seq_len=self.min_seq_len,
+            num_seqs_multiple_of=num_seqs_multiple_of,
+        )
+
+        return builder.bucket_by_length(
+            bucket_sizes,
+            selector=self.seq_len_col,
+            min_data_len=self.min_seq_len,
+            skip_below_min_examples=self.skip_below_min_examples,
+            skip_above_max_examples=True,
+        )
+
+
+def create_hf_reader(
+    dataset: Dataset,
+    gang: Gang,
+    converter: Callable[[Example], Seq2SeqBatch],
```

Comment:
I think instead of hard-coding to `Seq2SeqBatch`, making the batch type a generic type would be a lot more flexible (for an example see [here](https://github.com/facebookresearch/fairseq2/blob/main/src/fairseq2/datasets/data_reader.py#L23).